src/dependencies.ts

```diff
@@ -72,3 +75,52 @@ export function verifyDependencies(
   const failures = results.filter((r) => r.passed === false).length;
   return { failures, results };
 }
+
+export function findCyclesInDependencyGraph(graph: Graph): Cycles {
+  if (graph.edges().some((e) => !graph.isDirected(e))) {
+    throw new SfdxError('All edges must be directed');
+  }
+  return cyclicGraphDetection(graph);
+}
+
+export type Cycle = {
+  between: {
+    nodeA: string;
+    nodeB: string;
+  };
+  cycleAsPath: string[];
+};
+
+export type Cycles = Cycle[];
+
+function cyclicGraphDetection(graph: Graph): Cycles {
+  let p: string[] = [];
+  let previousDepth = 0;
+  const cycles: Cycles = [];
+  const g = graph;
+  dfs(graph, function (node, attr, depth) {
+    // starting with a new top level node
+    if (depth === 0) {
```

Comment:
Need to build a path to help detect cycles
depth of 0 - new path being searched - reset the path
depth <= previousDepth - previous call reached max depth for the path - set path to current path sliced from 0 up to depth.
Otherwise still searching current path